File: YASAudio/AudioEngine/YASAudioDeviceIONode.m
//
//  YASAudioDeviceIONode.m
//  Copyright (c) 2015 Yuki Yasoshima.
//

#import "YASAudioDeviceIONode.h"
#import "YASAudioDevice.h"
#import "YASAudioNode+Internal.h"
#import "YASAudioGraph.h"
#import "YASAudioConnection+Internal.h"
#import "YASAudioDeviceIO.h"
#import "YASAudioTapNode.h"
#import "YASAudioChannelRoute.h"
#import "YASAudioData+Internal.h"
#import "YASAudioFormat.h"
#import "YASAudioTime.h"
#import "YASWeakSupport.h"
#import "YASMacros.h"
#import "NSNumber+YASAudio.h"
#import "NSException+YASAudio.h"
#import "YASAudioUtility.h"

@interface YASAudioDeviceIONodeCore : YASAudioNodeCore

@property (nonatomic, strong) NSSet *outputChannelRoutes;
@property (nonatomic, strong) NSSet *inputChannelRoutes;

@end

@implementation YASAudioDeviceIONodeCore

- (void)dealloc
{
    YASRelease(_outputChannelRoutes);
    YASRelease(_inputChannelRoutes);

    _outputChannelRoutes = nil;
    _inputChannelRoutes = nil;

    YASSuperDealloc;
}

- (NSArray *)outputChannelRoutesWithSourceBus:(NSNumber *)sourceBus format:(YASAudioFormat *)format
{
    if (!_outputChannelRoutes) {
        return [YASAudioChannelRoute defaultChannelRoutesWithBus:sourceBus.uint32Value format:format];
    }

    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K == %@", YASAudioSourceBusKey, sourceBus];
    NSSortDescriptor *chDesc = [NSSortDescriptor sortDescriptorWithKey:YASAudioSourceChannelKey ascending:YES];
    NSSet *filteredSet = [_outputChannelRoutes filteredSetUsingPredicate:predicate];
    NSArray *sortedAssignments = [filteredSet sortedArrayUsingDescriptors:@[chDesc]];

    BOOL result = NO;

    if (sortedAssignments.count == format.channelCount) {
        result = YES;
        for (UInt32 i = 0; i < sortedAssignments.count; i++) {
            YASAudioChannelRoute *route = sortedAssignments[i];
            if (route.sourceChannel != i) {
                result = NO;
                break;
            }
        }
    }

    if (!result) {
        YASRaiseWithReason(([NSString stringWithFormat:@"%s - Invalid channel routes.", __PRETTY_FUNCTION__]));
        return nil;
    }

    return sortedAssignments;
}

- (NSArray *)inputChannelRoutesWithDestinationBus:(NSNumber *)destinationBus format:(YASAudioFormat *)format
{
    if (!_inputChannelRoutes) {
        return [YASAudioChannelRoute defaultChannelRoutesWithBus:destinationBus.uint32Value format:format];
    }

    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K == %@", YASAudioDestinationBusKey, destinationBus];
    NSSortDescriptor *chDesc = [NSSortDescriptor sortDescriptorWithKey:YASAudioDestinationChannelKey ascending:YES];
    NSSet *filteredSet = [_inputChannelRoutes filteredSetUsingPredicate:predicate];
    NSArray *sortedAssignments = [filteredSet sortedArrayUsingDescriptors:@[chDesc]];

    BOOL result = NO;

    if (sortedAssignments.count == format.channelCount) {
        result = YES;
        for (UInt32 i = 0; i < sortedAssignments.count; i++) {
            YASAudioChannelRoute *route = sortedAssignments[i];
            if (route.destinationChannel != i) {
                result = NO;
                break;
            }
        }
    }

    if (!result) {
        YASRaiseWithReason(([NSString stringWithFormat:@"%s - Invalid channel routes.", __PRETTY_FUNCTION__]));
        return nil;
    }

    return sortedAssignments;
}

@end

@interface YASAudioDeviceIONode ()

@property (nonatomic, strong) YASAudioDeviceIO *audioDeviceIO;
@property (atomic, strong) YASWeakContainer *graphContainer;
@property (nonatomic, strong) YASAudioDeviceIONodeCore *nodeCoreOnRender;

@end

@implementation YASAudioDeviceIONode {
    YASAudioDevice *_audioDevice;
}

- (instancetype)init
{
    YASAudioDevice *device = [YASAudioDevice defaultOutputDevice];
    return [self initWithAudioDevice:device];
}

- (instancetype)initWithAudioDevice:(YASAudioDevice *)audioDevice
{
    self = [super init];
    if (self) {
        if (!audioDevice) {
            YASRaiseWithReason(([NSString
                stringWithFormat:@"%s - Argument is nil. audioDevice(%@)", __PRETTY_FUNCTION__, audioDevice]));
            YASRelease(self);
            return nil;
        }
        _audioDevice = YASRetain(audioDevice);
        [self updateNodeCore];
    }
    return self;
}

- (void)dealloc
{
    YASRelease(_nodeCoreOnRender);
    YASRelease(_audioDevice);
    YASRelease(_audioDeviceIO);
    YASRelease(_graphContainer);

    _nodeCoreOnRender = nil;
    _audioDevice = nil;
    _audioDeviceIO = nil;
    _graphContainer = nil;

    YASSuperDealloc;
}

- (void)setAudioDevice:(YASAudioDevice *)audioDevice
{
    if (![_audioDevice isEqualToAudioDevice:audioDevice]) {
        YASRelease(_audioDevice);
        _audioDevice = YASRetain(audioDevice);

        self.audioDeviceIO.audioDevice = audioDevice;
        [self _notifyDidChangeAudioDevice];
    }
}

- (YASAudioDevice *)audioDevice
{
    return _audioDevice;
}

- (void)addAudioDeviceIOToGraph:(YASAudioGraph *)graph
{
    if (!graph) {
        YASRaiseWithReason(
            ([NSString stringWithFormat:@"%s - Argument is nil. graph(%@)", __PRETTY_FUNCTION__, graph]));
        return;
    }

    if (_audioDeviceIO) {
        YASRaiseWithReason(([NSString stringWithFormat:@"%s - AudioDeviceIO is already added.", __PRETTY_FUNCTION__]));
        return;
    }

    self.graphContainer = graph.weakContainer;

    YASAudioDeviceIO *deviceIO = [[YASAudioDeviceIO alloc] initWithAudioDevice:self.audioDevice];
    self.audioDeviceIO = deviceIO;
    [graph addAudioDeviceIO:deviceIO];
    YASRelease(deviceIO);
}

- (void)removeAudioDeviceIOFromGraph
{
    if (!_graphContainer) {
        YASRaiseWithReason(([NSString
            stringWithFormat:@"%s - AudioGraph is nil. graphContainer(%@)", __PRETTY_FUNCTION__, _graphContainer]));
        return;
    }

    if (!_audioDeviceIO) {
        YASRaiseWithReason(([NSString stringWithFormat:@"%s - AudioDeviceIO is removed. audioDeviceIO(%@)",
                                                       __PRETTY_FUNCTION__, _audioDeviceIO]));
        return;
    }

    YASAudioGraph *graph = [self.graphContainer retainedObject];
    [graph removeAudioDeviceIO:_audioDeviceIO];
    YASRelease(graph);

    self.graphContainer = nil;
    self.audioDeviceIO = nil;
}

- (void)prepareParameters
{
}

- (BOOL)_validateConnections
{
    @autoreleasepool
    {
        YASAudioDeviceIO *audioDeviceIO = self.audioDeviceIO;

        if (audioDeviceIO) {
            NSDictionary *inputConnections = self.inputConnections;
            if (inputConnections.count > 0) {
                for (YASAudioConnection *connection in inputConnections.objectEnumerator) {
                    YASAudioFormat *connectionFormat = connection.format;
                    YASAudioFormat *deviceFormat = audioDeviceIO.audioDevice.outputFormat;
                    if (connectionFormat.bitDepthFormat != deviceFormat.bitDepthFormat ||
                        connectionFormat.sampleRate != deviceFormat.sampleRate || connectionFormat.isInterleaved) {
                        YASLog(@"%s - Output device io format is not match.", __PRETTY_FUNCTION__);
                        return NO;
                    }
                }
            }
            NSDictionary *outputConnections = self.outputConnections;
            if (outputConnections.count > 0) {
                for (YASAudioConnection *connection in outputConnections.objectEnumerator) {
                    YASAudioFormat *connectionFormat = connection.format;
                    YASAudioFormat *deviceFormat = audioDeviceIO.audioDevice.inputFormat;
                    if (connectionFormat.bitDepthFormat != deviceFormat.bitDepthFormat ||
                        connectionFormat.sampleRate != deviceFormat.sampleRate || connectionFormat.isInterleaved) {
                        YASLog(@"%s - Input device io format is not match.", __PRETTY_FUNCTION__);
                        return NO;
                    }
                }
            }
        }
    }

    return YES;
}

- (void)updateConnections
{
    @autoreleasepool
    {
        YASAudioDeviceIO *audioDeviceIO = self.audioDeviceIO;
        if (!audioDeviceIO) {
            return;
        }

        if (![self _validateConnections]) {
            audioDeviceIO.renderCallbackBlock = nil;
            return;
        }

        YASWeakContainer *nodeContainer = self.weakContainer;
        YASWeakContainer *deviceIOContainer = audioDeviceIO.weakContainer;

        audioDeviceIO.renderCallbackBlock = ^(YASAudioData *outData, YASAudioTime *when) {
            @autoreleasepool
            {
                YASAudioDeviceIONode *node = [nodeContainer autoreleasingObject];
                if (!node) {
                    return;
                }

                YASAudioDeviceIONodeCore *core = node.nodeCore;
                if (!core) {
                    return;
                }

                self.nodeCoreOnRender = core;

                if (outData) {
                    NSDictionary *connections = core.inputConnections;
                    for (YASAudioConnection *connection in connections.objectEnumerator) {
                        NSArray *channelRoutes =
                            [core outputChannelRoutesWithSourceBus:connection.sourceBus format:connection.format];
                        if (channelRoutes) {
                            YASAudioFormat *format = connection.format;
                            YASAudioData *renderData = [[YASAudioData alloc] initWithFormat:format
                                                                                       data:outData
                                                                        outputChannelRoutes:channelRoutes];

                            YASAudioNode *sourceNode = connection.sourceNode;
                            [sourceNode renderWithData:renderData bus:connection.sourceBus when:when];

                            YASRelease(renderData);
                        }
                    }
                }

                {
                    YASAudioDeviceIO *deviceIO = [deviceIOContainer retainedObject];
                    if (deviceIO) {
                        NSDictionary *connections = core.outputConnections;
                        for (YASAudioConnection *connection in connections.objectEnumerator) {
                            id destinationNode = connection.destinationNode;
                            if ([destinationNode isKindOfClass:[YASAudioInputTapNode class]]) {
                                YASAudioInputTapNode *inputTapNode = destinationNode;
                                YASAudioFormat *format = connection.format;

                                NSArray *channelRoutes =
                                    [core inputChannelRoutesWithDestinationBus:connection.destinationBus format:format];
                                if (channelRoutes) {
                                    YASAudioData *inputData = [deviceIO inputDataOnRender];
                                    YASAudioTime *inputTime = [deviceIO inputTimeOnRender];
                                    if (inputData && inputTime) {
                                        YASAudioData *renderData = [[YASAudioData alloc] initWithFormat:format
                                                                                                   data:inputData
                                                                                     inputChannelRoutes:channelRoutes];

                                        [inputTapNode renderWithData:renderData bus:@0 when:inputTime];

                                        YASRelease(renderData);
                                    }
                                }
                            }
                        }

                        YASRelease(deviceIO);
                    }
                }

                self.nodeCoreOnRender = nil;
            }
        };
    }
}

- (UInt32)inputBusCount
{
    return UINT32_MAX;
}

- (UInt32)outputBusCount
{
    return UINT32_MAX;
}

- (void)setOutputChannelAssignments:(NSSet *)routes
{
    if (_outputChannelRoutes != routes) {
        YASRelease(_outputChannelRoutes);
        _outputChannelRoutes = YASRetain(routes);

        [self updateNodeCore];
    }
}

- (void)setInputChannelAssignments:(NSSet *)routes
{
    if (_inputChannelRoutes != routes) {
        YASRelease(_inputChannelRoutes);
        _inputChannelRoutes = YASRetain(routes);

        [self updateNodeCore];
    }
}

#pragma mark Private

- (Class)nodeCoreClass
{
    return [YASAudioDeviceIONodeCore class];
}

- (id)newNodeCoreObject
{
    YASAudioDeviceIONodeCore *core = [super newNodeCoreObject];
    core.outputChannelRoutes = _outputChannelRoutes;
    core.inputChannelRoutes = _inputChannelRoutes;

    return core;
}

- (void)_notifyDidChangeAudioDevice
{
}

#pragma mark Render thread

- (void)renderWithData:(YASAudioData *)data bus:(NSNumber *)bus when:(YASAudioTime *)when
{
    [super renderWithData:data bus:bus when:when];

    @autoreleasepool
    {
        YASAudioDeviceIO *audioDeviceIO = self.audioDeviceIO;
        if (audioDeviceIO) {
            YASAudioFormat *format = data.format;
            YASAudioFormat *deviceFormat = audioDeviceIO.audioDevice.inputFormat;
            if (format.bitDepthFormat != deviceFormat.bitDepthFormat || format.isInterleaved) {
                YASRaiseWithReason(
                    ([NSString stringWithFormat:@"%s - Format is not match. dataFormat(%@) deviceFormat(%@)",
                                                __PRETTY_FUNCTION__, format, deviceFormat]));
                return;
            }

            YASAudioDeviceIONodeCore *nodeCore = self.nodeCoreOnRender;
            if (nodeCore) {
                NSArray *channelRoutes = [nodeCore inputChannelRoutesWithDestinationBus:bus format:format];
                if (channelRoutes) {
                    YASAudioData *inputData = [audioDeviceIO inputDataOnRender];
                    const AudioBufferList *inputAbl = inputData.audioBufferList;
                    const AudioBufferList *outputAbl = data.audioBufferList;
                    if (inputAbl && !YASAudioIsEqualAudioBufferListStructure(inputAbl, outputAbl)) {
                        YASAudioData *renderData = [[YASAudioData alloc] initWithFormat:format
                                                                                   data:inputData
                                                                     inputChannelRoutes:channelRoutes];

                        [data copyFlexiblyFromData:renderData];

                        YASRelease(renderData);
                    }
                }
            }
        }
    }
}

@end

File: audio/audio_engine/yas_audio_device_io_node.cpp
//
//  yas_audio_device_io_node.cpp
//

#include "yas_audio_device_io_node.h"

#if (TARGET_OS_MAC && !TARGET_OS_IPHONE)

#include <iostream>
#include "yas_audio_device.h"
#include "yas_audio_device_io.h"
#include "yas_audio_graph.h"
#include "yas_audio_tap_node.h"
#include "yas_audio_time.h"
#include "yas_result.h"

using namespace yas;

#pragma mark - audio::device_io_node::impl

struct yas::audio::device_io_node::impl : base::impl, manageable_device_io_node::impl {
    audio::node _node = {{.input_bus_count = 1, .output_bus_count = 1}};
    audio::node::observer_t _connections_observer;

    impl() : _core(std::make_unique<core>()) {
    }

    virtual ~impl() final = default;

    void prepare(device_io_node const &device_io_node, audio::device const &device) {
        set_device(device ?: device::default_output_device());

        auto weak_node = to_weak(device_io_node);

        _node.set_render_handler(
            [weak_node](audio::pcm_buffer &buffer, uint32_t const bus_idx, audio::time const &when) {
                if (auto device_io_node = weak_node.lock()) {
                    if (auto const &device_io = device_io_node.impl_ptr<impl>()->device_io()) {
                        auto &input_buffer = device_io.input_buffer_on_render();
                        if (input_buffer && input_buffer.format() == buffer.format()) {
                            buffer.copy_from(input_buffer);
                        }
                    }
                }
            });

        _connections_observer =
            _node.subject().make_observer(audio::node::method::update_connections, [weak_node](auto const &) {
                if (auto device_io_node = weak_node.lock()) {
                    device_io_node.impl_ptr<impl>()->_update_device_io_connections();
                }
            });
    }

    void add_device_io() override {
        _core->_device_io = audio::device_io{_core->device()};
    }

    void remove_device_io() override {
        _core->_device_io = nullptr;
    }

    audio::device_io &device_io() override {
        return _core->_device_io;
    }

    void set_device(audio::device const &device) {
        _core->set_device(device);
    }

    audio::device device() const {
        return _core->device();
    }

   private:
    struct core {
        audio::device_io _device_io = nullptr;

        void set_device(audio::device const &device) {
            _device = device;
            if (_device_io) {
                _device_io.set_device(device);
            }
        }

        audio::device device() const {
            return _device;
        }

       private:
        audio::device _device = nullptr;
    };

    std::unique_ptr<core> _core;

    void _update_device_io_connections() {
        auto &device_io = _core->_device_io;
        if (!device_io) {
            return;
        }

        if (!_validate_connections()) {
            device_io.set_render_callback(nullptr);
            return;
        }

        auto weak_node = to_weak(cast<device_io_node>());
        auto weak_device_io = to_weak(device_io);

        auto render_function = [weak_node, weak_device_io](auto args) {
            if (auto node = weak_node.lock()) {
                if (auto kernel = node.node().kernel()) {
                    if (args.output_buffer) {
                        auto const connections = kernel.input_connections();
                        if (connections.count(0) > 0) {
                            auto const &connection = connections.at(0);
                            if (auto source_node = connection.source_node()) {
                                if (connection.format() == source_node.output_format(connection.source_bus())) {
                                    source_node.render(args.output_buffer, connection.source_bus(), args.when);
                                }
                            }
                        }
                    }

                    if (auto const device_io = weak_device_io.lock()) {
                        auto const connections = kernel.output_connections();
                        if (connections.count(0) > 0) {
                            auto const &connection = connections.at(0);
                            if (auto destination_node = connection.destination_node()) {
                                if (destination_node.is_input_renderable()) {
                                    auto input_buffer = device_io.input_buffer_on_render();
                                    auto const &input_time = device_io.input_time_on_render();
                                    if (input_buffer && input_time) {
                                        if (connection.format() ==
                                            destination_node.input_format(connection.destination_bus())) {
                                            destination_node.render(input_buffer, 0, input_time);
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        };

        device_io.set_render_callback(std::move(render_function));
    }

    bool _validate_connections() {
        if (auto const &device_io = _core->_device_io) {
            auto &input_connections = _node.input_connections();
            if (input_connections.size() > 0) {
                auto const connections = lock_values(input_connections);
                if (connections.count(0) > 0) {
                    auto const &connection = connections.at(0);
                    auto const &connection_format = connection.format();
                    auto const &device_format = device_io.device().output_format();
                    if (connection_format != device_format) {
                        std::cout << __PRETTY_FUNCTION__ << " : output device io format is not match." << std::endl;
                        return false;
                    }
                }
            }

            auto &output_connections = _node.output_connections();
            if (output_connections.size() > 0) {
                auto const connections = lock_values(output_connections);
                if (connections.count(0) > 0) {
                    auto const &connection = connections.at(0);
                    auto const &connection_format = connection.format();
                    auto const &device_format = device_io.device().input_format();
                    if (connection_format != device_format) {
                        std::cout << __PRETTY_FUNCTION__ << " : input device io format is not match." << std::endl;
                        return false;
                    }
                }
            }
        }

        return true;
    }
};

#pragma mark - audio::device_io_node

audio::device_io_node::device_io_node() : device_io_node(audio::device(nullptr)) {
}

audio::device_io_node::device_io_node(std::nullptr_t) : base(nullptr) {
}

audio::device_io_node::device_io_node(audio::device const &device) : base(std::make_unique<impl>()) {
    impl_ptr<impl>()->prepare(*this, device);
}

audio::device_io_node::~device_io_node() = default;

void audio::device_io_node::set_device(audio::device const &device) {
    impl_ptr<impl>()->set_device(device);
}

audio::device audio::device_io_node::device() const {
    return impl_ptr<impl>()->device();
}

audio::node const &audio::device_io_node::node() const {
    return impl_ptr<impl>()->_node;
}

audio::node &audio::device_io_node::node() {
    return impl_ptr<impl>()->_node;
}

audio::manageable_device_io_node &audio::device_io_node::manageable() {
    if (!_manageable) {
        _manageable = audio::manageable_device_io_node{impl_ptr<manageable_device_io_node::impl>()};
    }
    return _manageable;
}

#endif
